subsys/net/l2/openthread/openthread.c

```diff
@@ -227,17 +239,75 @@ void ot_receive_handler(otMessage *aMessage, void *context)
 	otMessageFree(aMessage);
 }
 
+static struct openthread_context *ot_context_for_restarter;
+
+void ot_joiner_start_handler(otError error, void *context);
+
+static void ot_joiner_restarter(void *dummy)
+{
+	otError error = otJoinerStart(ot_context_for_restarter->instance,
+					joiner_pskd, NULL,
+					PACKAGE_NAME, OT_PLATFORM_INFO,
+					PACKAGE_VERSION, NULL,
+					&ot_joiner_start_handler,
+					ot_context_for_restarter);
+
+	if (error != OT_ERROR_NONE) {
+		NET_ERR("Failed to start joiner [%d]", error);
+	}
+}
+
+K_WORK_DEFINE(joiner_restart, ot_joiner_restarter);
+
+static void ot_joiner_restart_timer(void *dummy)
+{
+	k_work_submit(&joiner_restart);
+}
+
+K_TIMER_DEFINE(joiner_restart_timer, ot_joiner_restart_timer, NULL);
 void ot_joiner_start_handler(otError error, void *context)
 {
 	struct openthread_context *ot_context = context;
-
+	static u32_t restart_delay;
+	static u32_t join_attempts;
 	switch (error) {
 	case OT_ERROR_NONE:
 		NET_INFO("Join success");
 		otThreadSetEnabled(ot_context->instance, true);
 		break;
 	default:
 		NET_ERR("Join failed [%d]", error);
+		if (IS_ENABLED(CONFIG_OPENTHREAD_JOINER_CONTINUOUS)) {
+			join_attempts++;
+
+			if (CONFIG_OPENTHREAD_JOINER_REBOOT != 0) {
+				if (join_attempts >=
+					CONFIG_OPENTHREAD_JOINER_REBOOT) {
+					printf("Joining failed %d times ... "
```

Comment:
Please, let's not introduce a bulky `printf` to the final binary. We should use logger here, just as in other places. If you're not getting an output because of the reboot, there's a function for that:
https://docs.zephyrproject.org/latest/reference/logging/index.html#_CPPv49log_panicv